crates/noirc_frontend/src/ast/mod.rs

```diff
@@ -149,13 +165,27 @@ impl std::fmt::Display for UnresolvedTypeExpression {
 }
 
 impl UnresolvedType {
-    pub fn from_int_token(token: IntType) -> UnresolvedType {
-        use {IntType::*, UnresolvedType::Integer};
+    pub fn without_span(typ: UnresolvedTypeData) -> UnresolvedType {
+        UnresolvedType { typ, span: None }
+    }
+
+    pub fn unspecified() -> UnresolvedType {
+        UnresolvedType { typ: UnresolvedTypeData::Unspecified, span: None }
+    }
+}
+
+impl UnresolvedTypeData {
+    pub fn from_int_token(token: IntType) -> UnresolvedTypeData {
+        use {IntType::*, UnresolvedTypeData::Integer};
         match token {
             Signed(num_bits) => Integer(Signedness::Signed, num_bits),
             Unsigned(num_bits) => Integer(Signedness::Unsigned, num_bits),
         }
     }
+
+    pub fn with_span(&self, span: Span) -> UnresolvedType {
+        UnresolvedType { typ: self.clone(), span: Some(span) }
```

Comment:
```suggestion
    pub fn with_span(self, span: Span) -> UnresolvedType {
        UnresolvedType { typ: self, span: Some(span) }
```
If the method requires an owned self, the signature should reflect that. Looking at where this method was used it was causing a lot of unnecessary clones before.